docs/docs/recipes/20-gmd.md

```diff
@@ -0,0 +1,104 @@
+import OpenNew from '@site/static/img/open_new.svg';
+
+# Configuring Testify to run on Gradle managed device
+
+Starting from API levels
+27 [Android Gradle Plugin allows you to configure virtual test devices in your project's Gradle files <OpenNew />](https://developer.android.com/studio/test/gradle-managed-devices).
+This feature improves testing experience by delegating the task of starting, shutting down and
+managing emulator to the Gradle.
+
+When using Gradle managed device user must
+use [Test storage service <OpenNew />](https://developer.android.com/reference/androidx/test/services/storage/TestStorage)
+to save screenshots. Using this service Android Gradle Plugin would be able to save test output
+including screenshots and diffs to the `build` folder
+
+### Adding Gradle managed device configuration to build file
+
+To create new Gradle managed device add it's definition to the `build.gradle` file:
+
+```groovy
+
+android {
+    testOptions {
+        managedDevices {
+            devices {
+                NAME_OF_THE_DEVICE(com.android.build.api.dsl.ManagedVirtualDevice) {
+                    device = "Pixel 2"
+                    apiLevel = 30
+                    systemImageSource = "aosp"
+                }
+            }
+        }
+    }
+}
+
+```
+
+In order to receive screenshots after test execution enable `TestStorageService`:
+
+```groovy
+
+android {
+    defaultConfig {
+        testInstrumentationRunnerArguments useTestStorageService: "true"
+    }
+}
+
+dependencies {
+    androidTestUtil("androidx.test.services:test-services:1.4.2")
+}
+```
+
+With that changes applied couple of new Gradle tasks should be added to your project. Suppose, you
+named your device `tester`, then `testerCheck` and `testDebugAndroidTest` tasks should be added to
+your project.
+
+### Configuring the Testify library
+
+You also need to instruct Testify to use Test storage to store screenshots and diffs. This could be
+done using Testify plugin configuration:
+
+```groovy
+testify {
+    useTestStorage = true
+}
+```
+
+### Running screenshot tests
+
+To perform screenshot test verification run any of the new Gradle tasks, for
+example `./gradlew testerDebugAndroidTest`.
+
+Because we need to run specific Gradle task to execute our tests using Gradle managed device, we
+cannot use `screenshotPull` task. When execution completed go to the
+module's `build/outputs/managed_device_android_test_additional_output/tester` folder to get recorded
```

Comment:
This appears to be working pretty well, but I have noticed a couple of limitations that may require some refinement in follow-up PRs:

1. Testify supports various locales and screen orientations by encoding that information into the directory name (device key, eg. `30-1080x2220@440dp-en_US`). This information is lost in the GMD `managed_device_android_test_additional_output` directory. It would be beneficial to improve the file output for the Test Storage destination to preserve this information
2. Similar to above, it would be handy to preserve the device key as part of the `managed_device_android_test_additional_output` so that it would be easier to copy the screenshots to the proper `assets` directory
3. It would be really nice if the `screenshotPull` task were able to understand when a GMD device is in-use and correctly copy the files from `managed_device_android_test_additional_output` automatically. This should be possible by checking the state of the `useTestStorage` flag in `TestifySettings` -- similar to how we check the SDCard flag.

Reply:
1-2. I am not sure about that, need to experiment
3. Yes, sounds like a good imrovement

Reply:
Regarding 1-2 I think the following: when running GMD, screenshots are stored in the `managed_device_android_test_additional_output` directory in the folder named after the name of the device given in the `managedDevices` clojure in the `build.gradle` file.
So, we have two options:
- either suggest user to name device following naming convention `VERSION-WIDTHxHEIGHT@DENSITY-LOCALE`
- or run some script after gradle task to rename the folder ?

Reply:
For 3 point we need to read `managedDevices` clojure in the `build.gradle` to understand where take screenshots from. I think it's also better to keep for further improvement

Reply:
Yes, these are good ideas for future improvement -- don't need to address this now